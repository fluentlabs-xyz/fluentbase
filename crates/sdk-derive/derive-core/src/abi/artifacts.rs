use chrono::Local;
use convert_case::{Case, Casing};
use serde_json::Value;
use std::{collections::HashSet, fs, path::Path};

const INTERFACE_TEMPLATE: &str = r#"// SPDX-License-Identifier: MIT
// This file is auto-generated
// DO NOT EDIT THIS FILE MANUALLY!
// Source contract: {contract_name}
// Generated at: {timestamp}

pragma solidity ^0.8.0;

{imports}
interface I{contract_name} {
{structs}
{functions}
}
"#;

const STRUCT_TEMPLATE: &str = r#"    struct {name} {
{fields}
    }"#;

const FUNCTION_TEMPLATE: &str = r#"    function {name}({params}) external{mutability}{returns};"#;

#[derive(Debug)]
pub struct ContractArtifacts {
    pub base_name: String,
    pub abi_content: String,
    pub interface_content: String,
}

#[derive(Debug)]
struct SolidityFunction {
    name: String,
    params: Vec<SolidityParameter>,
    returns: Vec<SolidityParameter>,
    is_view: bool,
    is_payable: bool,
}

#[derive(Debug)]
struct SolidityParameter {
    name: String,
    type_name: String,
    location: Option<DataLocation>,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Copy)]
enum DataLocation {
    Memory,
    Calldata,
    Storage,
}

impl ContractArtifacts {
    pub fn new(name: &str, abi: &[Value]) -> Self {
        let generator = SolidityInterfaceGenerator::new(name);
        Self {
            base_name: name.to_case(Case::Pascal),
            abi_content: serde_json::to_string_pretty(abi).expect("Failed to serialize ABI"),
            interface_content: generator.generate_from_abi(abi),
        }
    }

    pub fn save_to_dir(&self, dir: &Path) -> std::io::Result<()> {
        let contract_dir = dir.join(&self.base_name);
        fs::create_dir_all(&contract_dir)?;

        fs::write(
            contract_dir.join(format!("{}.abi.json", self.base_name)),
            &self.abi_content,
        )?;

        fs::write(
            contract_dir.join(format!("I{}.sol", self.base_name)),
            &self.interface_content,
        )?;

        Ok(())
    }
}

#[derive(Debug, Clone)]
pub struct SolidityInterfaceGenerator {
    contract_name: String,
    imports: Vec<String>,
}

impl SolidityInterfaceGenerator {
    pub fn new(name: &str) -> Self {
        Self {
            contract_name: name.to_case(Case::Pascal),
            imports: Vec::new(),
        }
    }

    pub fn generate_from_abi(&self, abi: &[Value]) -> String {
        let imports = self
            .imports
            .iter()
            .map(|import| format!("import \"{}\";", import))
            .collect::<Vec<_>>()
            .join("\n");

        let structs = self.collect_and_generate_structs(abi);
        let functions = self.generate_functions(abi);

        INTERFACE_TEMPLATE
            .replace("{contract_name}", &self.contract_name)
            .replace(
                "{timestamp}",
                &Local::now().format("%Y-%m-%d %H:%M:%S").to_string(),
            )
            .replace("{imports}", &imports)
            .replace("{structs}", &structs)
            .replace("{functions}", &functions)
    }

    fn collect_and_generate_structs(&self, abi: &[Value]) -> String {
        let struct_names = self.collect_struct_names(abi);
        let struct_definitions: Vec<String> = struct_names
            .iter()
            .filter_map(|name| self.generate_struct_definition(name))
            .collect();

        if struct_definitions.is_empty() {
            String::new()
        } else {
            format!("{}\n", struct_definitions.join("\n\n"))
        }
    }

    fn collect_struct_names(&self, abi: &[Value]) -> HashSet<String> {
        let mut seen_structs = HashSet::new();

        for func in abi {
            if let Some(inputs) = func.get("inputs").and_then(Value::as_array) {
                Self::collect_struct_types(inputs, &mut seen_structs);
            }
            if let Some(outputs) = func.get("outputs").and_then(Value::as_array) {
                Self::collect_struct_types(outputs, &mut seen_structs);
            }
        }

        seen_structs
    }

    fn collect_struct_types(params: &[Value], seen_structs: &mut HashSet<String>) {
        for param in params {
            if param["type"] == "tuple" {
                if let Some(components) = param.get("components").and_then(Value::as_array) {
                    Self::collect_struct_types(components, seen_structs);

                    if let Some(name) = param
                        .get("internalType")
                        .and_then(Value::as_str)
                        .filter(|s| s.starts_with("struct "))
                    {
                        seen_structs.insert(name[7..].to_string());
                    }
                }
            }
        }
    }

    fn generate_struct_definition(&self, struct_name: &str) -> Option<String> {
        let out_dir = std::env::var("OUT_DIR").ok()?;
        let file_path = Path::new(&out_dir)
            .join("solidity_abi")
            .join(format!("{}.json", struct_name));

        let content = fs::read_to_string(file_path).ok()?;
        let def: Value = serde_json::from_str(&content).ok()?;
        let components = def.get("components")?.as_array()?;

        let fields = components
            .iter()
            .map(|component| {
                let name = component["name"].as_str().unwrap();
                let type_str = Self::get_type_string(component);
                format!("        {} {};", type_str, name)
            })
            .collect::<Vec<_>>()
            .join("\n");

        Some(
            STRUCT_TEMPLATE
                .replace("{name}", struct_name)
                .replace("{fields}", &fields),
        )
    }

    fn generate_functions(&self, abi: &[Value]) -> String {
        abi.iter()
            .filter_map(|func| self.parse_function(func))
            .map(|func| self.format_function(&func))
            .collect::<Vec<_>>()
            .join("\n")
    }

    fn parse_function(&self, func: &Value) -> Option<SolidityFunction> {
        let name = func.get("name")?.as_str()?.to_string();
        let inputs = func.get("inputs")?.as_array()?;
        let outputs = func.get("outputs")?.as_array()?;
        let state_mutability = func.get("stateMutability").and_then(Value::as_str);

        let params = inputs
            .iter()
            .map(|param| self.parse_parameter(param))
            .collect();

        let returns = outputs
            .iter()
            .map(|param| self.parse_parameter(param))
            .collect();

        Some(SolidityFunction {
            name,
            params,
            returns,
            is_view: state_mutability == Some("view"),
            is_payable: state_mutability == Some("payable"),
        })
    }

    fn parse_parameter(&self, param: &Value) -> SolidityParameter {
        let name = param["name"].as_str().unwrap_or("").to_string();
        let type_str = Self::get_type_string(param);
        let location = self.get_data_location(param);

        SolidityParameter {
            name,
            type_name: type_str,
            location,
        }
    }

    fn format_function(&self, func: &SolidityFunction) -> String {
        let params = func
            .params
            .iter()
            .map(|param| self.format_parameter(param, true))
            .collect::<Vec<_>>()
            .join(", ");

        let mutability = if func.is_payable {
            " payable"
        } else if func.is_view {
            " view"
        } else {
            ""
        };

        let returns = if func.returns.is_empty() {
            String::new()
        } else {
            let return_params = func
                .returns
                .iter()
                .map(|param| self.format_parameter(param, false))
                .collect::<Vec<_>>()
                .join(", ");
            format!(" returns ({})", return_params)
        };

        FUNCTION_TEMPLATE
            .replace("{name}", &func.name)
            .replace("{params}", &params)
            .replace("{mutability}", mutability)
            .replace("{returns}", &returns)
    }

    fn format_parameter(&self, param: &SolidityParameter, is_input: bool) -> String {
        let location = param.location.map_or(String::new(), |loc| {
            format!(
                " {}",
                match (is_input, loc) {
                    (true, DataLocation::Memory) => "memory",
                    (true, DataLocation::Calldata) => "calldata",
                    (false, _) => "memory",
                    (_, DataLocation::Storage) => "storage",
                }
            )
        });

        if param.name.is_empty() {
            format!("{}{}", param.type_name, location)
        } else {
            format!("{}{} {}", param.type_name, location, param.name)
        }
    }

    fn get_type_string(param: &Value) -> String {
        let type_str = param["type"].as_str().unwrap();

        if type_str == "tuple" {
            if let Some(internal_type) = param
                .get("internalType")
                .and_then(Value::as_str)
                .filter(|s| s.starts_with("struct "))
            {
                return internal_type[7..].to_string();
            }

            let components = param["components"].as_array().unwrap();
            let component_types = components
                .iter()
                .map(Self::get_type_string)
                .collect::<Vec<_>>();

            format!("({})", component_types.join(","))
        } else {
            type_str.to_string()
        }
    }

    fn get_data_location(&self, param: &Value) -> Option<DataLocation> {
        let type_str = param["type"].as_str().unwrap();
        match type_str {
            "string" | "bytes" => Some(DataLocation::Calldata),
            "tuple" => Some(DataLocation::Memory),
            _ if type_str.ends_with("[]") => Some(DataLocation::Memory),
            _ => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use serial_test::serial;
    use tempfile::TempDir;

    #[test]
    fn test_simple_function() {
        let abi = vec![json!({
            "name": "greeting",
            "type": "function",
            "stateMutability": "view",
            "inputs": [
                {"name": "name", "type": "string"}
            ],
            "outputs": [
                {"name": "result", "type": "string"}
            ]
        })];

        let generator = SolidityInterfaceGenerator::new("Greeter");
        let interface = generator.generate_from_abi(&abi);

        assert!(interface.contains("interface IGreeter {"));
        assert!(interface.contains(
            "function greeting(string calldata name) external view returns (string memory result);"
        ));
    }

    #[test]
    #[serial]
    fn test_struct_generation() {
        let temp_dir = TempDir::new().unwrap();
        std::env::set_var("OUT_DIR", temp_dir.path());

        let abi_dir = temp_dir.path().join("solidity_abi");
        fs::create_dir_all(&abi_dir).unwrap();

        let point_abi = json!({
            "type": "tuple",
            "components": [
                {"name": "x", "type": "uint64"},
                {"name": "y", "type": "uint256"}
            ]
        });

        fs::write(
            abi_dir.join("Point.json"),
            serde_json::to_string_pretty(&point_abi).unwrap(),
        )
        .unwrap();

        let abi = vec![json!({
            "name": "setPoint",
            "type": "function",
            "inputs": [{
                "name": "point",
                "type": "tuple",
                "internalType": "struct Point",
                "components": [
                    {"name": "x", "type": "uint64"},
                    {"name": "y", "type": "uint256"}
                ]
            }],
            "outputs": []
        })];

        let generator = SolidityInterfaceGenerator::new("PointManager");
        let interface = generator.generate_from_abi(&abi);

        assert!(interface.contains("struct Point {"));
        assert!(interface.contains("uint64 x;"));
        assert!(interface.contains("uint256 y;"));
        assert!(interface.contains("function setPoint(Point memory point) external;"));
    }

    #[test]
    #[serial]
    fn test_nested_structs() {
        let temp_dir = TempDir::new().unwrap();
        std::env::set_var("OUT_DIR", temp_dir.path());

        let abi_dir = temp_dir.path().join("solidity_abi");
        fs::create_dir_all(&abi_dir).unwrap();

        let point_abi = json!({
            "type": "tuple",
            "components": [
                {"name": "x", "type": "uint64"},
                {"name": "y", "type": "uint256"}
            ]
        });

        let complex_point_abi = json!({
            "type": "tuple",
            "components": [
                {
                    "name": "point",
                    "type": "tuple",
                    "internalType": "struct Point",
                    "components": [
                        {"name": "x", "type": "uint64"},
                        {"name": "y", "type": "uint256"}
                    ]
                },
                {"name": "description", "type": "string"},
                {"name": "active", "type": "bool"}
            ]
        });

        fs::write(
            abi_dir.join("Point.json"),
            serde_json::to_string_pretty(&point_abi).unwrap(),
        )
        .unwrap();

        fs::write(
            abi_dir.join("ComplexPoint.json"),
            serde_json::to_string_pretty(&complex_point_abi).unwrap(),
        )
        .unwrap();

        let abi = vec![json!({
            "name": "processPoint",
            "type": "function",
            "inputs": [{
                "name": "complexPoint",
                "type": "tuple",
                "internalType": "struct ComplexPoint",
                "components": [
                    {
                        "name": "point",
                        "type": "tuple",
                        "internalType": "struct Point",
                        "components": [
                            {"name": "x", "type": "uint64"},
                            {"name": "y", "type": "uint256"}
                        ]
                    },
                    {"name": "description", "type": "string"},
                    {"name": "active", "type": "bool"}
                ]
            }],
            "outputs": [{"name": "success", "type": "bool"}]
        })];

        let generator = SolidityInterfaceGenerator::new("PointProcessor");
        let interface = generator.generate_from_abi(&abi);

        assert!(interface.contains("struct Point {"));
        assert!(interface.contains("struct ComplexPoint {"));

        assert!(interface.contains("uint64 x;"));
        assert!(interface.contains("uint256 y;"));
        assert!(interface.contains("Point point;"));
        assert!(interface.contains("string description;"));
        assert!(interface.contains("bool active;"));

        assert!(interface.contains("function processPoint(ComplexPoint memory complexPoint) external returns (bool success);"));
    }

    #[test]
    fn test_array_types() {
        let abi = vec![json!({
            "name": "processArrays",
            "type": "function",
            "inputs": [
                {"name": "numbers", "type": "uint256[]"},
                {"name": "names", "type": "string[]"}
            ],
            "outputs": [
                {"name": "result", "type": "bool[]"}
            ]
        })];

        let generator = SolidityInterfaceGenerator::new("ArrayProcessor");
        let interface = generator.generate_from_abi(&abi);

        assert!(interface.contains(
            "function processArrays(uint256[] memory numbers, string[] memory names) external returns (bool[] memory result);"
        ));
    }

    #[test]
    fn test_anonymous_tuple() {
        let abi = vec![json!({
            "name": "processTuple",
            "type": "function",
            "inputs": [{
                "name": "data",
                "type": "tuple",
                "components": [
                    {"name": "value", "type": "uint256"},
                    {"name": "flag", "type": "bool"}
                ]
            }],
            "outputs": []
        })];

        let generator = SolidityInterfaceGenerator::new("TupleProcessor");
        let interface = generator.generate_from_abi(&abi);

        assert!(interface.contains("function processTuple((uint256,bool) memory data) external;"));
    }

    #[test]
    #[serial]
    fn test_artifacts_saving() {
        let temp_dir = TempDir::new().unwrap();

        let abi = vec![json!({
            "name": "test",
            "type": "function",
            "inputs": [],
            "outputs": []
        })];

        let artifacts = ContractArtifacts::new("TestContract", &abi);
        artifacts.save_to_dir(temp_dir.path()).unwrap();

        let contract_dir = temp_dir.path().join("TestContract");
        assert!(contract_dir.exists());
        assert!(contract_dir.join("TestContract.abi.json").exists());
        assert!(contract_dir.join("ITestContract.sol").exists());
    }
}
