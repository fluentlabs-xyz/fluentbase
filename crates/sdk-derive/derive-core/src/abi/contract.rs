use crate::abi::function::FunctionABI;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Represents a stored ABI definition with metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContractABI {
    pub id: String,
    pub functions: Vec<FunctionABI>,

    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub metadata: HashMap<String, String>,
}

#[allow(dead_code)]
impl ContractABI {
    pub fn new(id: impl Into<String>) -> Self {
        Self {
            id: id.into(),
            functions: Vec::new(),
            metadata: HashMap::new(),
        }
    }

    #[must_use]
    pub fn with_function(mut self, function: FunctionABI) -> Self {
        self.functions.push(function);
        self
    }

    #[must_use]
    pub fn with_functions(mut self, functions: Vec<FunctionABI>) -> Self {
        self.functions.extend(functions);
        self
    }

    pub fn with_metadata(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.metadata.insert(key.into(), value.into());
        self
    }
    #[must_use]
    pub fn path(&self) -> String {
        self.id.replace("::", "/")
    }
}
