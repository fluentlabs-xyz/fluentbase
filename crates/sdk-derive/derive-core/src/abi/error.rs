#[derive(Debug, thiserror::Error)]
pub enum ABIError {
    #[error("Invalid type conversion: {0}")]
    TypeConversion(String),

    #[error("Serialization error: {0}")]
    Serialization(String),

    #[error("Deserialization error: {0}")]
    Deserialization(String),

    #[error("Unsupported type: {0}")]
    UnsupportedType(String),

    #[error("Unsupported pattern: {0}")]
    UnsupportedPattern(String),

    #[error("Syntax error: {0}")]
    Syntax(String),

    #[error("Internal error: {0}")]
    Internal(String),

    #[error("Config error: {0}")]
    Config(String),
}

pub type ABIResult<T> = std::result::Result<T, ABIError>;
