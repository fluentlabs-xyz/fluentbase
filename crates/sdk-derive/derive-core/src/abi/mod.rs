//! ABI module provides functionality for working with Solidity ABI, focusing on function signatures
//! parsing and type conversions.
//!
//! # Core components
//!
//! * `SolType` - represents Solidity types, parses Rust types into their Solidity equivalents
//! * `FunctionABI` - represents Solidity function definitions
//! * `RustToSol` - converts Rust types to their Solidity equivalents using registry
//!
//! # Function ID Generation
//!
//! Module can generate function signatures and IDs from Rust code in two modes:
//!
//! 1. Registry mode (enabled via feature flag):
//!    - All structure definitions are loaded into registry at startup
//!    - When constructing `FunctionABI`, registry is used to resolve types
//!    - Provides complete type information for complex structures
//!
//! 2. Direct mode (default):
//!    - No registry preloading
//!    - Types are converted directly during `FunctionABI` construction
//!    - Suitable for simple cases without complex nested structures
//!
//! ```rust, ignore
//! use crate::abi::{FunctionABI, RustToSol};
//!
//! // Parse Rust function signature
//! let sig: syn::Signature = parse_quote! {
//!     fn transfer(amount: u64, recipient: String) -> String
//! };
//!
//! // Create registry (optional)
//! // check artifacts for structure definitions
//!
//!  let config = ArtifactsRegistryConfig::new("OUT_DIR").with_mirror(".artifacts");
//! let registry = ArtifactsRegistry::new(config)?;
//!
//!
//! // Convert to FunctionABI
//! let abi = FunctionABI::from_syn(&sig, registry)?;
//!
//! // Get function ID (first 4 bytes of keccak256 hash)
//! let function_id = abi.function_id();
//!
//! // Get canonical function signature (e.g. "transfer(uint256,string)")
//! let signature = abi.signature();
//! ```
//!
//! # Constraints
//!
//! - Types must implement `SolidityABI` derive macro
//! - Generic types are not supported
//! - Module path used for type identification in registry
//! - Only non-generic Rust types can be converted to Solidity types
//! - Complex types must be registered in the registry when using Registry mode
pub mod contract;
pub mod error;
pub mod function;
pub mod parameter;
pub mod types;
