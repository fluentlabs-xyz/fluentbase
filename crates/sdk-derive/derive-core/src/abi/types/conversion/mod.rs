mod rust_to_sol;
mod sol_to_rust;

pub use rust_to_sol::{rust_to_sol, ConversionError};
pub use sol_to_rust::sol_to_rust;
