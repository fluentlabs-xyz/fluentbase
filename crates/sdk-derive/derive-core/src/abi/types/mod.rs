mod conversion;
mod sol;

pub use conversion::*;
pub use sol::SolType;
