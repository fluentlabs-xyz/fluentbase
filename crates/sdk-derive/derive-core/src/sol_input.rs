use crate::abi::types::{convert_solidity_type, sol_to_rust};
use alloy_sol_macro_input::{SolInput, SolInputKind};
use convert_case::{Case, Casing};
use proc_macro2::{Span, TokenStream};
use quote::{format_ident, quote};
use syn::{Ident, Type};
use syn_solidity::{
    visit::{visit_file, Visit},
    File,
    Item,
    ItemFunction,
    ItemStruct,
    Mutability,
    Spanned,
    VariableDeclaration,
};

/// A visitor that collects Solidity functions and structs
#[derive(Default)]
struct Collector<'a> {
    functions: Vec<&'a ItemFunction>,
    structs: Vec<&'a ItemStruct>,
}

impl<'a> Visit<'a> for Collector<'a> {
    fn visit_item_function(&mut self, func: &'a ItemFunction) {
        self.functions.push(func);
    }
    fn visit_item_struct(&mut self, s: &'a ItemStruct) {
        self.structs.push(s);
    }
}

/// Converts Solidity input to Rust trait token stream
///
/// # Arguments
///
/// * `input` - The Solidity input to convert
///
/// # Returns
///
/// A TokenStream representing the generated Rust trait
pub fn to_rust_trait(input: SolInput) -> syn::Result<TokenStream> {
    let (structs, trait_name, trait_fns) = convert_sol_to_rust(input)?;

    // Generate the final output for trait
    Ok(quote! {
        #(#structs)*
        pub trait #trait_name {
            #(#trait_fns)*
        }
    })
}

/// Converts Solidity input to Rust client trait with solidity mode
///
/// # Arguments
///
/// * `input` - The Solidity input to convert
///
/// # Returns
///
/// A TokenStream representing the generated Rust client trait
pub fn to_sol_client(input: SolInput) -> syn::Result<TokenStream> {
    let (structs, trait_name, trait_fns) = convert_sol_to_rust(input)?;

    // Generate the final output for client trait with attribute
    Ok(quote! {
        #(#structs)*
        #[client(mode = "solidity")]
        pub trait #trait_name {
            #(#trait_fns)*
        }
    })
}

/// Converts Solidity input to Rust code components
///
/// # Arguments
///
/// * `input` - The Solidity input to convert
///
/// # Returns
///
/// A tuple of (structs, trait_name, trait_methods) to be assembled
fn convert_sol_to_rust(
    input: SolInput,
) -> syn::Result<(Vec<TokenStream>, Ident, Vec<TokenStream>)> {
    // Get the Solidity file from the input
    let file = match input.kind {
        SolInputKind::Sol(sol_file) => sol_file,
        SolInputKind::Type(_) => {
            return Err(syn::Error::new(
                Span::call_site(),
                "Expected Solidity interface or contract, not type",
            ));
        } /* TODO(d1r1): Handle JSON ABI
           * #[cfg(feature = "json")]
           * SolInputKind::Json(_, _) => {
           *     return Err(syn::Error::new(
           *         Span::call_site(),
           *         "JSON ABI not supported in this macro",
           *     ));
           * } */
    };

    let mut visitor = Collector::default();
    visit_file(&mut visitor, &file);

    let trait_name = derive_trait_name(&file)?;

    let structs = visitor
        .structs
        .iter()
        .map(|s| sol_struct_to_rust_tokens(s))
        .collect::<syn::Result<Vec<_>>>()?;

    let trait_fns = visitor
        .functions
        .iter()
        .filter_map(|func| sol_fn_to_trait_method(func).ok())
        .filter(|tokens| !tokens.is_empty())
        .collect::<Vec<_>>();

    Ok((structs, trait_name, trait_fns))
}

/// Derives a trait name from the Solidity file
///
/// # Arguments
///
/// * `file` - The Solidity file
///
/// # Returns
///
/// The derived trait name as an Ident or an error if no contract was found
fn derive_trait_name(file: &File) -> syn::Result<Ident> {
    for item in &file.items {
        if let Item::Contract(contract) = item {
            let name = contract.name.to_string().to_case(Case::Pascal);

            // If it's an interface, use the name directly
            if contract.kind.is_interface() {
                return Ok(format_ident!("{}", name));
            }

            // If it's a contract, prefix with 'I'
            if contract.kind.is_contract() {
                return Ok(format_ident!("I{}", name));
            }
        }
    }

    // Return error if no contract or interface is found
    Err(syn::Error::new(
        Span::call_site(),
        "No Solidity contract or interface found in input",
    ))
}

/// Determines the method receiver type based on mutability
///
/// # Arguments
///
/// * `func` - The Solidity function
///
/// # Returns
///
/// A TokenStream representing the receiver (&self or &mut self)
fn determine_method_receiver(func: &ItemFunction) -> TokenStream {
    match func.attributes.mutability() {
        Some(Mutability::View(_)) | Some(Mutability::Pure(_)) => quote! { &self },
        _ => quote! { &mut self },
    }
}

/// Converts a Solidity struct to Rust struct tokens
///
/// # Arguments
///
/// * `sol_struct` - The Solidity struct
///
/// # Returns
///
/// A TokenStream representing the generated Rust struct
fn sol_struct_to_rust_tokens(sol_struct: &ItemStruct) -> syn::Result<TokenStream> {
    let name = &sol_struct.name;

    // Convert all fields
    let fields = sol_struct
        .fields
        .iter()
        .map(|field| {
            // Get field name or use _ as default
            let field_name = field
                .name
                .as_ref()
                .map(|n| Ident::new(&n.to_string(), n.span()))
                .unwrap_or_else(|| Ident::new("_", field.ty.span()));

            // Convert Solidity type to Rust type
            let sol_ty = convert_solidity_type(&field.ty)?;
            let rust_ty: Type = sol_to_rust(&sol_ty).map_err(|e| {
                syn::Error::new(field.ty.span(), format!("Struct field type error: {e}"))
            })?;

            Ok(quote! { pub #field_name: #rust_ty })
        })
        .collect::<syn::Result<Vec<_>>>()?;

    // Generate the struct definition
    Ok(quote! {
        #[derive(::fluentbase_sdk::codec::Codec, Debug, Clone, PartialEq, Eq)]
        pub struct #name {
            #(#fields),*
        }
    })
}

/// Converts a Solidity function to a Rust trait method
///
/// # Arguments
///
/// * `func` - The Solidity function
///
/// # Returns
///
/// A TokenStream representing the generated trait method
fn sol_fn_to_trait_method(func: &ItemFunction) -> syn::Result<TokenStream> {
    // Skip functions without a name or special functions
    let Some(name) = &func.name else {
        return Ok(quote! {});
    };

    if name == "fallback" || name == "receive" {
        return Ok(quote! {});
    }

    // Convert function name to snake_case
    let fn_name = format_ident!("{}", name.to_string().to_case(Case::Snake));
    let receiver = determine_method_receiver(func);

    // Generate function parameters
    let args = func
        .parameters
        .iter()
        .enumerate()
        .filter_map(|(i, param)| sol_param_to_tokens(i, param).ok())
        .collect::<Vec<_>>();

    // Generate function return type
    let ret = sol_return_to_tokens(func)?;

    // Generate the function signature
    Ok(quote! {
        fn #fn_name(#receiver #(, #args)*) #ret;
    })
}

/// Converts a Solidity function parameter to Rust tokens
///
/// # Arguments
///
/// * `index` - The parameter index
/// * `param` - The parameter declaration
///
/// # Returns
///
/// A TokenStream representing the parameter
fn sol_param_to_tokens(index: usize, param: &VariableDeclaration) -> syn::Result<TokenStream> {
    // Get parameter name or generate one
    let name_str = param
        .name
        .as_ref()
        .map(|n| n.to_string().to_case(Case::Snake))
        .unwrap_or_else(|| format!("_param{index}"));

    let name_ident = format_ident!("{}", name_str);

    // Convert Solidity type to Rust type
    let sol_ty = convert_solidity_type(&param.ty)?;
    let rust_ty = sol_to_rust(&sol_ty)
        .map_err(|e| syn::Error::new(param.ty.span(), format!("Cannot convert param type: {e}")))?;

    Ok(quote! { #name_ident: #rust_ty })
}

/// Converts a Solidity function return to Rust tokens
///
/// # Arguments
///
/// * `func` - The Solidity function
///
/// # Returns
///
/// A TokenStream representing the return type
fn sol_return_to_tokens(func: &ItemFunction) -> syn::Result<TokenStream> {
    // If there are no returns, return an empty token stream
    let Some(returns) = &func.returns else {
        return Ok(quote! {});
    };

    let return_params = &returns.returns;

    // If there's only one return parameter, use it directly
    if return_params.len() == 1 {
        let sol_ty = convert_solidity_type(&return_params[0].ty)?;
        let rust_ty = sol_to_rust(&sol_ty).map_err(|e| {
            syn::Error::new(
                return_params[0].ty.span(),
                format!("Return type error: {e}"),
            )
        })?;

        return Ok(quote! { -> #rust_ty });
    }

    // For multiple return parameters, create a tuple
    let rust_types = return_params
        .iter()
        .map(|param| {
            let sol_ty = convert_solidity_type(&param.ty)?;
            sol_to_rust(&sol_ty).map(|ty| quote! { #ty }).map_err(|e| {
                syn::Error::new(param.ty.span(), format!("Tuple return type error: {e}"))
            })
        })
        .collect::<syn::Result<Vec<_>>>()?;

    Ok(quote! { -> (#(#rust_types),*) })
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use prettyplease;
    use syn::parse_str;

    #[test]
    fn test_sol_struct_to_rust_tokens() {
        let sol = r#"
            struct User {
                string name;
                uint256 age;
            }
        "#;

        let item: syn_solidity::ItemStruct = parse_str(sol).unwrap();
        let tokens = sol_struct_to_rust_tokens(&item).unwrap();
        let file = syn::parse_file(&tokens.to_string()).unwrap();
        let formatted = prettyplease::unparse(&file);

        assert_snapshot!("sol_struct_to_rust_tokens", formatted);
    }

    #[test]
    fn test_no_contract_error() {
        let solidity_code = r#"
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

library SomeLibrary {
    function doSomething() public pure returns (uint256) {
        return 42;
    }
}
"#;
        let input: alloy_sol_macro_input::SolInput = parse_str(solidity_code).unwrap();

        let result = to_rust_trait(input);
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("No Solidity contract or interface found"));
    }

    #[test]
    fn test_to_rust_trait_nested_struct() {
        let solidity_code = r#"
        // SPDX-License-Identifier: MIT
        pragma solidity ^0.8.0;

        interface IProgram {
            struct Inner {
                uint256 x;
            }

            struct Outer {
                Inner inner;
                string note;
            }

            function ping(Outer calldata input) external view returns (Outer calldata);
        }
    "#;
        let input: alloy_sol_macro_input::SolInput = parse_str(solidity_code).unwrap();

        let generated = to_rust_trait(input).unwrap();
        let parsed = syn::parse_file(&generated.to_string()).unwrap();
        let formatted = prettyplease::unparse(&parsed);

        assert_snapshot!("sol_to_rust_trait_nested_struct", formatted);
    }

    #[test]
    fn test_to_sol_client_nested_struct() {
        let solidity_code = r#"
        // SPDX-License-Identifier: MIT
        pragma solidity ^0.8.0;

        interface IProgram {
            struct Inner {
                uint256 x;
            }

            struct Outer {
                Inner inner;
                string note;
            }

            function ping(Outer calldata input) external view returns (Outer calldata);
        }
    "#;
        let input: alloy_sol_macro_input::SolInput = parse_str(solidity_code).unwrap();

        let generated = to_sol_client(input).unwrap();
        let parsed = syn::parse_file(&generated.to_string()).unwrap();
        let formatted = prettyplease::unparse(&parsed);

        assert_snapshot!("sol_to_sol_client_nested_struct", formatted);
    }
}
