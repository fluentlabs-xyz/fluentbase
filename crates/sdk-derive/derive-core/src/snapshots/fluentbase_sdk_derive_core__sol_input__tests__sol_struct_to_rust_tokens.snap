---
source: crates/sdk-derive/derive-core/src/sol_input.rs
expression: formatted
---
#[derive(::fluentbase_sdk::codec::Codec, Debug, Clone, PartialEq, Eq)]
pub struct User {
    pub name: String,
    pub age: U256,
}
